# Rala
Layout module for raw de novo DNA assembly of long uncorrected reads.

## Description
Rala is intended as a standalone layout module to assemble raw reads generated by third generation sequencing. It is a combination of two approaches, chimeric and repeat detection proposed by [HINGE](https://github.com/HingeAssembler/HINGE) and assembly graph simplification as proposed by [Miniasm](https://github.com/lh3/miniasm). In preprocessing, with the help of coverage graphs, chimeric reads and overlaps between unbridged repeat reads are removed (see figures bellow). Chimeric reads have a deep coverage pit while repeats are characterized by coverage hills. After preprocessing is done, an assembly graph is built and simplified with transitive reduction, trimming, bubble popping and, as a last resort, heuristics which untangle junctions in graph.
Rala takes as input two files: raw reads in FASTQ format and overlaps between them produced by [Minimap](https://github.com/lh3/minimap) in PAF or MHAP format. If using PAF, reads should be preprocessed with *'misc/fastq_formatter.py'* which replaces read names with identifiers, **before** running Minimap.
Output is a set of contigs in FASTA format.

![alt text](http://github.com/rvaser/rala/master/misc/chimeric_read.png)
![alt text](http://github.com/rvaser/rala/master/misc/false_overlap.png)

## Dependencies
1. gcc 4.8+

## Installation
To install Rala run to following commands:

    git clone --recursive https://github.com/rvaser/rala.git rala
    cd rala/
    make

After running make, an executable named rala will appear in the current directory.

**Troubleshooting**

If you accidentally left out '--recursive' from git clone, run the following commands before running 'make':

    git submodule init
    git submodule update

## Usage

Run Rala with the following command:

    ./rala 0 <reads file> <overlaps file> <0 if MHAp, 1 if PAF>

## Contact information

For additional information, help and bug reports please send an email to: robert.vaser@fer.hr.

## Acknowledgement

This work has been supported in part by Croatian Science Foundation under the project UIP-11-2013-7353.
