# Rala
Layout module for raw de novo DNA assembly of long uncorrected reads.

## Description
Rala is intended as a standalone layout module to assemble raw reads generated by third generation sequencing. It was inspired by ideas used in [HINGE](https://github.com/HingeAssembler/HINGE) for preprocessing and in [Miniasm](https://github.com/lh3/miniasm) for assembly graph simplification. With the help of coverage graphs, chimeric reads are detected and cut while overlaps between unbridged repeat reads are removed from the dataset (see figures bellow). After preprocessing is done, an assembly graph is built and simplified with transitive reduction, trimming, bubble popping and, as a last resort, heuristics which untangle leftover junctions in graph.

Rala takes as input two files: raw reads in FASTQ format and overlaps between them produced by [Minimap](https://github.com/lh3/minimap) in PAF or MHAP format. If using PAF, reads should be preprocessed with *'misc/fastq_formatter.py'* which replaces read names with identifiers, **before** running Minimap. Output is a set of contigs in FASTA format.

*Coverage graph of a chimeric read*
![](misc/chimeric_read.png)

*Coverage graphs of repeat reads which overlap is removed*
![](misc/false_overlap.png)


## Dependencies
1. gcc 4.8+

## Installation
To install Rala run to following commands:

    git clone --recursive https://github.com/rvaser/rala.git rala
    cd rala/
    make

After running make, an executable named rala will appear in the current directory.

**Troubleshooting**

If you accidentally left out '--recursive' from git clone, run the following commands before running 'make':

    git submodule init
    git submodule update

## Usage

Run Rala with the following command:

    ./rala 0 <reads file> <overlaps file> <0 if MHAP, 1 if PAF>

## Contact information

For additional information, help and bug reports please send an email to: robert.vaser@fer.hr.

## Acknowledgement

This work has been supported in part by Croatian Science Foundation under the project UIP-11-2013-7353.
